### PR TITLE
Clean slate with Winnow 2

### PR DESCRIPTION
**Pull request by dhilst**
Continuing the rewrite started by @tertsdiepraam  https://github.com/uutils/parse_datetime/pull/61

* The functions in `lib.rs` keep the same signature
* I kept the tests in `lib.rs` too all they are passing 
* 90% of the parsing was written by @tertsdiepraam (check #61)
* Added timestamp with `@\d+` format
* Added timezone abbreviation names
* Added 8 digits dates `MMDDhhmm`
* Using [winnow](https://github.com/winnow-rs/winnow)

### CODE REVIEW

**Comment by dhilst** [comment]
@tertsdiepraam I squashed your commits into a single commit, I hope you don't mind

**Comment by tertsdiepraam** [comment]
Looks cool! A few initial comments to send you in the right direction. One of the reasons for this rewrite is that chrono wasn't cutting it for adding all the items together. Instead we should  add all the items together ourselves to really match GNU. It'll require a lot of experimentation, but it'll make the compatibility much better. I think most of the corner cases were around the ends of months. I believe GNU had some overflow behaviour there that chrono didn't. So I recommend first creating a custom datetime struct to manipulate and then mapping that to chrono.

Squashing the commit is absolutely fine by the way. Thanks for continuing this!

File: src/lib.rs [hunk 1]
```diff
@@ -381,7 +183,12 @@ mod tests {
 
         #[test]
         fn invalid_offset_format() {
-            let invalid_offsets = vec!["+0700", "UTC+2", "Z-1", "UTC+01005"];
+            let invalid_offsets = vec![
+                // "+0700",     // this is not invalid
+                // "UTC+2",     // this is not invalid
+                // "Z-1",       // this is not invalid, means UTC-1
+                "UTC+01005", // this is
+            ];
```
Comment (dhilst) [inline src/lib.rs:191]: These cases were failing and I commented on them because I don't know why they were considered invalid before. The code parses `UTC+2` as two time zones and sums them up. Since UTC equals 0, `UTC+2` is the same as `+2`. I don't know if this behavior is expected.

File: Cargo.toml [hunk 2]
```diff
@@ -1,13 +1,16 @@
 [package]
 name = "parse_datetime"
-description = "parsing human-readable time strings and converting them to a DateTime"
-version = "0.6.0"
+description = "parsing human-readable relative time strings and converting them to a Duration"
+version = "0.7.0"
```
Comment (sylvestre) [inline Cargo.toml:4]: the api didn't change?

Comment (dhilst) [reply Cargo.toml:4]: Oh, this was before recreating the parse_datetime function, I'm restoring it to 6

Comment (dhilst) [reply Cargo.toml:4]: Okay, I setting this to `0.7.0`. The API didn't change, for one minute I found that `0.6.1` was the right version, but this is not a simple bugfix so I think `0.7.0` is the correct choice.

Comment (sylvestre) [reply Cargo.toml:4]: this is fine, i was just curious :)

**Comment by sylvestre** [comment]
please let us know when it is ready to be reviewed :)

**Comment by dhilst** [comment]
> please let us know when it is ready to be reviewed :)

it is ready for review (:

File: src/items/date.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+pub fn year(input: &mut &str) -> PResult<u32> {
+    s(alt((
+        take(4usize).try_map(|x: &str| x.parse()),
+        take(3usize).try_map(|x: &str| x.parse()),
+        take(2usize).try_map(|x: &str| x.parse()).map(
+            |x: u32| {
+                if x <= 68 {
+                    x + 2000
+                } else {
+                    x + 1900
```
Comment (sylvestre) [inline src/items/date.rs:108]: seems that it isn't covered by a test, maybe a add test for it?

Comment (dhilst) [reply src/items/date.rs:108]: I added a test for the if (:

**Comment by sylvestre** [comment]
a test failed:
```

---- tests::iso_8601::test_t_sep stdout ----
error: test failed, to rerun pass `--lib`
thread 'tests::iso_8601::test_t_sep' panicked at src/lib.rs:128:13:
assertion `left == right` failed
  left: 2021-02-15T06:37:47-05:00
 right: 2021-02-15T06:37:47Z
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

**Comment by dhilst** [comment]
> a test failed:
> 
> ```
> 
> ---- tests::iso_8601::test_t_sep stdout ----
> error: test failed, to rerun pass `--lib`
> thread 'tests::iso_8601::test_t_sep' panicked at src/lib.rs:128:13:
> assertion `left == right` failed
>   left: 2021-02-15T06:37:47-05:00
>  right: 2021-02-15T06:37:47Z
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

I saw this in the CI, but I couldn't reproduce it here, for some reason it is passing to me, any ideas on how to reproduce?
image

**Comment by dhilst** [comment]
> > a test failed:
> > ```
> > 
> > ---- tests::iso_8601::test_t_sep stdout ----
> > error: test failed, to rerun pass `--lib`
> > thread 'tests::iso_8601::test_t_sep' panicked at src/lib.rs:128:13:
> > assertion `left == right` failed
> >   left: 2021-02-15T06:37:47-05:00
> >  right: 2021-02-15T06:37:47Z
> > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > ```
> 
> I saw this in the CI, but I couldn't reproduce it here, for some reason it is passing to me, any ideas on how to reproduce? image

Okay, I fixed the test. There was a `.fixed_offset()` call missing. Without this, it prints Z instead of the offset and the test fails

**Comment by dhilst** [comment]
The codecov is failing, there are some cases where it is checking if the test code is covered, should I fix them all?, or is there some configuration I'm missing!? codecov is kind of new to me 
image

**Comment by tertsdiepraam** [comment]
Codecov can be wrong sometimes. It's nice to have good coverage, but it might be alright.

File: src/items/time.rs [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+    fn from(
+        Offset {
+            negative,
+            hours,
+            minutes,
+        }: Offset,
+    ) -> Self {
+        let secs = hours * 3600 + minutes * 60;
+
+        eprintln!("------------------> {hours} {minutes} {secs}");
```
Comment (tertsdiepraam) [inline src/items/time.rs:100]: This should probably be removed. The commented out `eprintln`'s as well.

File: src/items/time.rs [hunk 5]
```diff
@@ -0,0 +1,806 @@
...
+
+        return Ok(newtz);
+    };
+
+    Ok(tz)
+}
+
+/// Timezone list extracted from:
+///   https://www.timeanddate.com/time/zones/
+fn tzname_to_offset(input: &str) -> PResult<Offset> {
```
Comment (tertsdiepraam) [inline src/items/time.rs:297]: There are crates like `chrono-tz` that can do stuff like this and should (hopefully) be updated. That might be a bit more convenient. We could accept this in the meantime though.

File: src/lib.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/lib.rs:252]: Sure! Feel free to delete them!

File: src/items/mod.rs [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                let day = day.into();
+
+                while beginning_of_day.weekday() != day {
+                    beginning_of_day += chrono::Duration::days(1);
+                }
+
+                d = beginning_of_day
+            }
+            Item::Relative(relative::Relative::Years(x)) => d = d.with_year(x).unwrap_or(d),
+            Item::Relative(relative::Relative::Months(x)) => {
```
Comment (tertsdiepraam) [inline src/items/mod.rs:273]: Could you write some tests around the handling of months? I'm not sure it matches GNU right now. I'm especially interested in cases like `28 Feb + 1 month` or `29 Feb + 1 year` and weird edge cases like that. I think I recorded some of those cases, but I forgot where unfortunately.

**Review by tertsdiepraam — COMMENTED** [review]
Very cool! I don't think anything really blocks this. I think there are some edge cases where it's not compatible but it would be so much better than the current version that I think that's alright.

File: src/items/mod.rs (hunk 6 above)
Comment (dhilst) [reply src/items/mod.rs:268]: Added some tests in `lib.rs` named `test_month`

Comment (tertsdiepraam) [reply src/items/mod.rs:268]: I'm sorry, I was confused before, I meant `2024-01-31 + 1 month`. GNU will give March 2, where `chrono` will yield February 29. Here's a test case:
```rust
#[test]
fn month_overflow() {
    env::set_var("TZ", "UTC");
    assert_eq!(
        parse_datetime("2024-01-31 + 1 month")
            .unwrap()
            .format("%+")
            .to_string(),
        "2024-03-02T00:00:00+00:00",
    )
}
```

**Comment by tertsdiepraam** [comment]
Hi! I was looking at this again and I found an issue with how you're currently using `chrono`. Here's my theory:
1. It's currently 2024-09-11 and September is a month with 30 days.
2. I give as input: `2024-01-31`
3. While setting the `Date` item, you first change the day, then the month and then the year.
4. At each of those, the validity is checked.
5. That means that you adjust the date first to `2024-09-31`, then `2024-01-31`.
6. However, `2024-09-31` is an invalid date and `chrono` will therefore yield and error even though I have entered a valid date.

We need to find a way to construct the times such that this never happens. Just changing the order won't cut it because that can still lead to invalid states. `chrono` might be more of a hindrance than a help here, unfortunately.

**Comment by dhilst** [comment]
> Hi! I was looking at this again and I found an issue with how you're currently using `chrono`. Here's my theory:
> 
>     1. It's currently 2024-09-11 and September is a month with 30 days.
> 
>     2. I give as input: `2024-01-31`
> 
>     3. While setting the `Date` item, you first change the day, then the month and then the year.
> 
>     4. At each of those, the validity is checked.
> 
>     5. That means that you adjust the date first to `2024-09-31`, then `2024-01-31`.
> 
>     6. However, `2024-09-31` is an invalid date and `chrono` will therefore yield and error even though I have entered a valid date.
> 
> 
> We need to find a way to construct the times such that this never happens. Just changing the order won't cut it because that can still lead to invalid states. `chrono` might be more of a hindrance than a help here, unfortunately.

Hi Terts, 

I introduced a function to create dates from naive dates instead of using `with_` method chains.

About the month behavior, if I understood correctly GNU gets the number of the days in the current month and adds it to the date, I implemented this behavior added your test and one more.

**Comment by dhilst** [comment]
@tertsdiepraam I was looking into the fuzzy testing. I added the code so it generates fuzzy dates with some sanity and caught some problems, I will look into them next.

File: fuzz/fuzz_targets/parse_datetime.rs [hunk 7]
```diff
@@ -1,8 +1,98 @@
...
+    //     .arg(format!("+{fmt}"))
+    //     .output()
+    //     .expect("date command failed");
+    // io::stdout().write_all(&output.stdout).unwrap();
+    // io::stdout().write_all(&output.stderr).unwrap();
+    // output.stdout.pop(); // remove trailing \n
+    // assert_eq!(
+    //     String::from_utf8(output.stdout).unwrap(),
+    //     d.format(fmt).to_string()
+    // );
```
Comment (dhilst) [inline fuzz/fuzz_targets/parse_datetime.rs:97]: I was testing something like this to determine its compatibility with GNU. The test run GNU `date` in and compare the outputs

Comment (tertsdiepraam) [reply fuzz/fuzz_targets/parse_datetime.rs:97]: We could definitely do that with fuzzing! We should still rely on regular tests as well though.

File: src/lib.rs [hunk 8]
```diff
@@ -529,4 +400,113 @@ mod tests {
...
+
+            // chrono rejects this specific date
+            assert!(NaiveDate::from_ymd_opt(262144, 3, 10).is_none());
+            assert!(chrono::DateTime::parse_from_str(input, FMT).is_err());
+            // the parsing works, but hydration fails
+            assert!(items::parse(&mut input.to_string().as_str()).is_ok());
+            assert!(parse_datetime(input).is_err());
+            // GNU date works
+            assert_eq!(make_gnu_date(input, FMT), input);
+        }
```
Comment (dhilst) [inline src/lib.rs:491]: @tertsdiepraam I found a case where GNU accepts the date, we can parse it and chrono rejects it, more precisely [chrono::NaiveDate::from_ymd_opt](https://docs.rs/chrono/latest/chrono/naive/struct.NaiveDate.html#method.from_ymd_opt) returns None for this date. I don't know what direction to take from there, chrono keeps getting in the way :sweat_smile: ideas? Maybe having a gnu-compatible datetime layer, but that will be a lot of work

Comment (dhilst) [reply src/lib.rs]: Another problem is with formatting too. _chrono_ puts a `+` in front of the year if the year is `>9999`. This breaks some tests:

image
image

**Comment by sylvestre** [comment]
sorry but some tests are failing:


```
failures:
    items::date::tests::test_year
    items::date::tests::with_year
    items::tests::date_and_time
    tests::offsets::invalid_offset_format
    tests::test_relative::chrono_date
    tests::test_relative::gnu_compat
    tests::test_relative::test_month
```

and needs some rebasing

**Comment by dhilst** [comment]
> sorry but some tests are failing:
> 
> ```
> failures:
>     items::date::tests::test_year
>     items::date::tests::with_year
>     items::tests::date_and_time
>     tests::offsets::invalid_offset_format
>     tests::test_relative::chrono_date
>     tests::test_relative::gnu_compat
>     tests::test_relative::test_month
> ```
> 
> and needs some rebasing

I will rebase it

But right now I have a problem with Chrono, I'm looking into alternatives like https://time-rs.github.io/book/how-to/create-dates.html

**Comment by dhilst** [comment]
Hello, I updated the PR, some key notes:

* I found that if we use only years in the range 0 to 9999, Chrono and GNU behavior coincide. I update the PR with such a fix, the fuzz tests for example are generating years in this range. 
* GNU accepts years in the range 0 to 2147485547
* Chrono accepts years in the range 0 to 262144
* But when using years >9999 the chrono add a '+' the output format where GNU doesn't

Let me know what you think, if this is too bad, or if it is okay for now. I guess that to achieve 100%
GNU compatibility we'll need to do our own date math. 

BTW I was playing with GNU date and I found this behavior

```
TZ=UTC date -d '0000-01-01 00:00 utc+00001' '+%Y'
-001
```

Here `-001` is the year, it looks like it is subtracting 1 second from the date (which is 0)
and it underflows to -1, I guess this is a bug, here is again all the output

```
Z=UTC date -d '0000-01-01 00:00 utc+00001' 
Fri Dec 31 11:59:00 PM UTC -001
```

File: Cargo.toml [hunk 9]
```diff
@@ -1,13 +1,26 @@
...
+winnow = { version="0.5.34"}
+
+[dev-dependencies]
+anyhow = "1.0.86"
+#winnow = { version="0.5.34", features = ["debug"] }
+
+[features]
+debug = ["winnow/debug"]
+
+[[bin]]
```
Comment (sylvestre) [inline Cargo.toml:22]: should be done in a different PR, no?

File: fuzz/Cargo.toml [hunk 10]
```diff
@@ -15,6 +15,9 @@ chrono = { version="0.4", default-features=false, features=["std", "alloc", "clo
 [dependencies.parse_datetime]
 path = "../"
 
+[features]
```
Comment (sylvestre) [inline fuzz/Cargo.toml:18]: same, why in this PR ?

File: Cargo.toml (hunk 9 above)
Comment (dhilst) [reply Cargo.toml:22]: Yes! I added to compare some dates with GNU, I'm removing this

**Comment by tertsdiepraam** [comment]
Succeeded by https://github.com/uutils/parse_datetime/pull/137

**Comment by dhilst** [comment]
> Succeeded by #137, which includes these commits. Thanks @dhilst!

I saw this yesterday. I'm glad that the work was useful. Thanks for all the help and reviews!! :rocket: :rocket: :rocket:
